lib/vector-core/src/event/array.rs

```diff
@@ -137,42 +137,29 @@ pub enum EventArray {
 }
 
 impl EventArray {
-    /// Call the given update function over each `LogEvent` in this array.
-    pub fn for_each_log(&mut self, update: impl FnMut(&mut LogEvent)) {
-        if let Self::Logs(logs) = self {
-            logs.iter_mut().for_each(update);
-        }
-    }
-
-    /// Call the given update function over each `Metric` in this array.
-    pub fn for_each_metric(&mut self, update: impl FnMut(&mut Metric)) {
-        if let Self::Metrics(metrics) = self {
-            metrics.iter_mut().for_each(update);
-        }
-    }
-
-    /// Run the given update function over each `Trace` in this array.
-    pub fn for_each_trace(&mut self, update: impl FnMut(&mut TraceEvent)) {
-        if let Self::Traces(traces) = self {
-            traces.iter_mut().for_each(update);
+    /// Iterate over references to this array's events.
+    pub fn iter_events(&self) -> impl Iterator<Item = EventRef> {
+        match self {
+            Self::Logs(array) => EventArrayIter::Logs(array.iter()),
+            Self::Metrics(array) => EventArrayIter::Metrics(array.iter()),
+            Self::Traces(array) => EventArrayIter::Traces(array.iter()),
         }
     }
 
-    /// Call the given update function over each event in this array.
-    pub fn for_each_event(&mut self, mut update: impl FnMut(EventMutRef<'_>)) {
+    /// Iterate over mutable references to this array's events.
+    pub fn iter_events_mut(&mut self) -> impl Iterator<Item = EventMutRef> {
         match self {
-            Self::Logs(array) => array.iter_mut().for_each(|log| update(log.into())),
-            Self::Metrics(array) => array.iter_mut().for_each(|metric| update(metric.into())),
-            Self::Traces(array) => array.iter_mut().for_each(|trace| update(trace.into())),
+            Self::Logs(array) => EventArrayIterMut::Logs(array.iter_mut()),
+            Self::Metrics(array) => EventArrayIterMut::Metrics(array.iter_mut()),
+            Self::Traces(array) => EventArrayIterMut::Traces(array.iter_mut()),
         }
     }
 
-    /// Iterate over this array's events.
-    pub fn iter_events(&self) -> impl Iterator<Item = EventRef> {
+    /// Iterate over references to the logs in this array.
+    pub fn iter_logs_mut(&mut self) -> impl Iterator<Item = &mut LogEvent> {
```

Comment:
Is the lack of similar `iter_metrics|traces_mut()` just because we haven't had a need yet?

Reply:
Exactly, yes.